src/geoparquet.js

```diff
@@ -0,0 +1,192 @@
+
+  const geoString = key_value_metadata?.find(kv => kv.key === 'geo')?.value
+  if (geoString === undefined) {
+    throw new Error('No geo field')
+  }
+
+  /** @type {unknown} */
+  let geo
+  try {
+    geo = JSON.parse(geoString)
+  } catch (_) {
+    throw new Error('Invalid GeoParquet metadata: geo field is not valid JSON')
+  }
+
+  if (
+    !geo ||
+    typeof geo !== 'object' ||
+    Array.isArray(geo)
+  ) {
+    throw new Error('Invalid GeoParquet metadata: not an object')
+  }
+
+  /* Version */
+  if (
+    !('version' in geo) ||
+    typeof geo.version !== 'string'
+  ) {
+    throw new Error('Invalid GeoParquet metadata: missing or invalid version')
+  }
+  if (geo.version !== '1.0.0' && geo.version !== '1.1.0' && geo.version !== '1.2.0-dev') {
+    throw new Error(`Unsupported GeoParquet version: ${geo.version}`)
+  }
+  const { version } = geo
+
+  /* Columns */
+  if (
+    !('columns' in geo) ||
+    !geo.columns ||
+    typeof geo.columns !== 'object' ||
+    Array.isArray(geo.columns)
+  ) {
+    throw new Error('Invalid GeoParquet metadata: missing or invalid columns')
+  }
+
+  /** @type GeoParquet['columns'] */
+  const columns = {}
+  for (const [columnName, metadata] of Object.entries(geo.columns)) {
+    columns[columnName] = parseColumn({ columnName, metadata, columnNames, version })
+  }
+
+  /* Primary column */
+  if (
+    !('primary_column' in geo) ||
+    typeof geo.primary_column !== 'string'
+  ) {
+    throw new Error('Invalid GeoParquet metadata: missing or invalid primary_column')
+  }
+  if (!(geo.primary_column in columns)) {
+    throw new Error(`Invalid GeoParquet metadata: primary_column "${geo.primary_column}" does not exist in columns`)
+  }
+  const { primary_column } = geo
+
+  return {
+    version,
+    primary_column,
+    columns,
+  }
+}
+
+/**
+ * Parse GeoParquet column metadata
+ * @param {object} options
+ * @param {string} options.columnName
+ * @param {unknown} options.metadata
+ * @param {string[]} options.columnNames
+ * @param {GeoParquet['version']} options.version
+ * @returns {GeoParquetColumn}
+ */
+export function parseColumn({ columnName, metadata, columnNames, version }) {
+  if (!columnNames.includes(columnName)) {
+    throw new Error(`Invalid GeoParquet metadata: no BYTE_ARRAY column called "${columnName}" in file`)
+  }
+
+  if (!metadata || typeof metadata !== 'object' || Array.isArray(metadata)) {
+    throw new Error(`Invalid GeoParquet metadata: column "${columnName}" is not an object`)
+  }
+
+  /* Encoding */
+  if (!('encoding' in metadata)) {
+    throw new Error(`Invalid GeoParquet metadata: column "${columnName}" missing encoding`)
+  }
+  if (
+    typeof metadata.encoding !== 'string' ||
+    !['WKB', ...unsupportedEncodings].includes(metadata.encoding)
+  ) {
+    throw new Error(`Invalid GeoParquet metadata: column "${columnName}" invalid encoding "${metadata.encoding}"`)
+  }
```

Comment:
aggressive validation, I would prefer that we broadly assume that files are valid and only check if we really need to. (applies to a bunch of this file)

Reply:
Yes, I saw it was clearly not the way uswd in the rest of the lib. I'll fix that.